File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/Bar.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class Bar {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    List<Object> bEdges = List.of();
    List<Object> aEdges = List.of();
    

    static class BarCleaner implements Runnable {

        MemorySegment segment;
        BarCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.Bar_destroy(this.segment);
        }
    }

    Bar() {}
    Bar(MemorySegment handle, List<Object> selfEdges, List<Object> bEdges, List<Object> aEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        this.bEdges = bEdges;
        this.aEdges = aEdges;
        

    }
    
    
    public Foo foo() {
        
        
        var nativeVal = somelib_h.Bar_foo(internal);
        
        List<Object> selfEdges = List.of(this);
        
        
        
        List<Object> aEdges = List.of(this);
        var returnVal = new Foo(nativeVal, selfEdges, aEdges);
        return returnVal;
                
    }
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/Bar.kt
package dev.diplomattest.somelib;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface BarLib: Library {
    fun Bar_destroy(handle: Pointer)
    fun Bar_foo(handle: Pointer): Pointer
}

class Bar internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal val bEdges: List<Any?>,
    internal val aEdges: List<Any?>,
)  {

    internal class BarCleaner(val handle: Pointer, val lib: BarLib) : Runnable {
        override fun run() {
            lib.Bar_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<BarLib> = BarLib::class.java
        internal val lib: BarLib = Native.load("diplomat_feature_tests", libClass)
    }
    
    fun foo(): Foo {
        // This lifetime edge depends on lifetimes: 'b, 'a
        val bEdges: MutableList<Any> = mutableListOf(this);
        // This lifetime edge depends on lifetimes: 'a
        val aEdges: MutableList<Any> = mutableListOf(this);
        
        val returnVal = lib.Bar_foo(handle);
        val selfEdges: List<Any> = listOf(this)
        val handle = returnVal 
        val returnOpaque = Foo(handle, selfEdges, aEdges)
        return returnOpaque
    }

}